### PR TITLE
Add docs build job

### PR DESCRIPTION
**Pull request by AyodeAwe**
The PR adds a `docs_build` process to the PR and Build workflows for this repository. The generated docs are synced to s3 for only the build workflows.

### CODE REVIEW

File: dependencies.yaml [hunk 1]
```diff
@@ -131,18 +137,21 @@ dependencies:
...
-          - sphinx-markdown-tables
+          - ipython
+          - numpydoc
+          - pip
+          - pip:
+            - breathe
+            - pydata_sphinx_theme
+            - sphinx_copybutton
+            - sphinx_markdown_tables
+          - recommonmark
```
Comment (ajschmidt8) [inline dependencies.yaml:154]: `breathe` is on this list twice.

also, the `pip` method of installing packages should be reserved for packages that aren't available on Anaconda.org.

It looks like all of these `pip` packages are currently available on Anaconda.org:

- https://anaconda.org/conda-forge/pydata-sphinx-theme
- https://anaconda.org/conda-forge/breathe
- https://anaconda.org/conda-forge/sphinx-copybutton
- https://anaconda.org/conda-forge/sphinx-markdown-tables

Therefore this `pip` section should be removed and its dependencies should be moved up a level.

**Comment by cjnolet** [comment]
@AyodeAwe it looks like we got past the depenendencies issue but now there's an issue of it not finding the docs index (should *should* have been built during the C++ libraft).

File: ci/build_docs.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  libraft-headers \
+  libraft-nn \
+  pylibraft \
+  raft-dask
+
+
+# Build CPP docs
+rapids-logger "Build CPP docs"
+pushd docs
+sphinx-build -b dirhtml source build -W
```
Comment (cjnolet) [inline ci/build_docs.sh:43]: RAFT uses a plugin called `breathe` for sphinx, which means cmake also gets involved in the docs build process and the python docs build is combined w/ the c++ docs build. Is there any reason we can't use `./build.sh docs` here?

**Review by cjnolet — CHANGES_REQUESTED** [review]

**Review by cjnolet — APPROVED** [review]
LGTM!

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]
This PR will need a few changes.

Blocking the merge until I have time for a proper review.

**Comment by cjnolet** [comment]
Sounds good @ajschmidt8 i was just very happy to see the docs seemed to build successfully in the logs.

**Comment by ajschmidt8** [comment]
To do:

- address these warnings: https://github.com/rapidsai/raft/actions/runs/4118640487/jobs/7112560748#step:6:721

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -52,6 +52,17 @@ jobs:
       branch: ${{ inputs.branch }}
       date: ${{ inputs.date }}
       sha: ${{ inputs.sha }}
+  docs-build:
+    if: ${{ startsWith(github.ref, 'refs/heads/branch-') }}
+    needs: python-build
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@cuda-118
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:59]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

File: .github/workflows/pr.yaml [hunk 4]
```diff
@@ -51,6 +52,16 @@ jobs:
     uses: rapidsai/shared-action-workflows/.github/workflows/conda-python-tests.yaml@branch-23.04
     with:
       build_type: pull-request
+  docs-build:
+    needs: conda-python-build
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@cuda-118
```
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:58]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

**Review by ajschmidt8 — APPROVED** [review]

**Pull request merged**